### ISSUE TITLE
Analyzer should add a diagnostic for import of opted out libraries from opted in libraries 

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/sdk#44063]
Based on an offline discussion with @leafpetersen @mit-mit @jonasfj @stereotype441 @natebosch and myself, we think that the analyzer should add at least a hint level diagnostic for any imports of opted out libraries from opted in libraries.

This is part of the broader strategy around `pub outdated --mode=null-safety` and `pub get`. We are going to remove the warnings from `pub get` for opted in packages, since pub does not have the logic to know what packages are actually imported. 

Pub will also now trust that a package is opted in to null safety purely based on its _default_ language version, and not require all libraries in the package to be opted in.

To counterbalance that, we would like to add this analyzer hint so that any user who does end up importing an opted out file from an opted in file will be notified right away by the analyzer.

This will also help with package scoring, as it will reduce the pub score of any package doing an out of order migration.

**Comment by scheglov** [comment]
https://dart-review.googlesource.com/c/sdk/+/170441

**Comment by scheglov** [comment]
There are [diagnostics](https://ci.chromium.org/p/dart/builders/try/flutter-analyze-try/12176) in Flutter.
Was this expected?

**Comment by jakemac53** [comment]
cc @yjbanov It looks like there are some web_ui violations of this lint, see https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket.appspot.com/8864529598122711632/+/steps/analyze_flutter_engine/0/stdout

For the most part the test deps look like they need to be updated to the null safe versions.

**Comment by jakemac53** [comment]
> Was this expected?

It was not completely unexpected, but we _want_ to know about these anyways as they do represent problems.

**Comment by scheglov** [comment]
And there are (many) [failing targets](https://test.corp.google.com/ui#id=OCL:340711341:BASE:340711432:1604523543881:ca539115) in google3 presubmit.

**Comment by jakemac53** [comment]
I will start looking at fixing the test_core issues

**Comment by jakemac53** [comment]
Note that this has revealed several issues with test_core where we could have been broken easily by other packages migrating so this is working great.

**Comment by leafpetersen** [comment]
cc @davidmorgan for how we want to manage this internally.

**Comment by jakemac53** [comment]
I need to land the test_core fixes and then we can get a better picture, I think in general we have not been allowing this already.

**Comment by jakemac53** [comment]
I have landed the test_core fixes internally, and started a rerun of the internal test failures

**Comment by jakemac53** [comment]
Definitely more things to work though now, I will continue working on this tomorrow. We could disable the hint for analysis internally (don't need anything from the analyzer afaik to do this).

**Comment by jakemac53** [comment]
Did a bit of investigation with adding analyzer ignores, this causes problems with the angular generator. I think we need to actually set the language version on these bootstrap files to match the file they are bootstrapping.

There is also what looks like probably a fairly long tail of other code generators and things that have issues.

To me it seems like we would want to just ignore this lint internally (for builds at least), although we should get it enabled there over time as well.

**Comment by davidmorgan** [comment]
Internally we have the `null_safety_migration_order` lint. Unusually for a lint, it's enforced on third party code too. But, we had to add a few exceptions; `test_core` was one :)

I think we'll just turn off the hint internally, it's useful that we can precisely control the lint. e.g. we're looking at special casing some things w.r.t. Angular. So, feel free to go ahead with the hint, we'll need to disable for google3 as it lands (or before, if the hint code is already known) @iinozemtsev

The lint has a few additional features to the hint, FYI:

 - We have an annotation `@blocksNullSafetyMigration`. By using this annotation a library can migrate to null safety but still block migration of its users. We're using this for Angular today since we're not ready to migrate users yet.
 - The lint does not fire if a legacy library is imported using `show` and all the shown symbols are defined in null safe libraries. This is an idea to help unblock migrations, we haven't used it yet.

One last suggestion, maybe link in the hint message to a doc explaining _why_ migration order matters?

Thanks everyone :)

**Comment by davidmorgan** [comment]
One more thought Jake--re: generators--are generators going to see the hint and complain about it? We don't have a way to turn off a particular hint in all the generators. If that's the case then this will be harder to land than I thought.

**Comment by jakemac53** [comment]
> One more thought Jake--re: generators--are generators going to see the hint and complain about it?

At least by default generators only complain about syntax errors. I don't _think_ this will be a problem but I am not totally sure. I didn't see any errors like this in the set of failures I was looking at but there were thousands of failures so I didn't look at a very representative sample.

**Comment by davidmorgan** [comment]
SG.

**Comment by scheglov** [comment]
I opened two PRs to ignore existing violations in flutter/flutter and flutter/engine.
https://github.com/flutter/flutter/pull/69904
https://github.com/flutter/engine/pull/22326

**Comment by franklinyow** [comment]
@scheglov any update?

**Comment by scheglov** [comment]
The previous PRs landed, but a new violation was added to Flutter while these PRs were waiting.
I have another PR to fix this: https://github.com/flutter/flutter/pull/70107, waiting for review.

**Comment by scheglov** [comment]
The previous PR landed, but yet another violation was added in between.
https://github.com/flutter/flutter/pull/70115

**Comment by jakemac53** [comment]
@scheglov I would just land your PR now - and alert the flutter roller sherrif about the issue so they can help resolve any issues that crop up before the roll happens (we will inevitably keep getting these)

**Comment by scheglov** [comment]
Landed the PR and CL.

**Comment by leafpetersen** [comment]
Cherry pick filed [here](https://github.com/dart-lang/sdk/issues/44182).
